[2단계 - 로또(수동)] 디우(김동규) 미션 제출합니다.

tco0427 opened the pull request:
안녕하세요 닉! 디우입니다..ㅎㅎ 이번 피드백도 잘 부탁드립니다!

---

2단계를 구현하며 가장 크게 고인되었던 부분은 `수동`이라는 기능이 추가되면서 책임을 어느쪽에 부여하는게 적절한가였습니다..
우선 수동과 자동 로또 구매 개수를 계산하고 검증하는 책임은 LottoPurchaseMoney 쪽에 부여하였는데요. 이러한 과정에서 calculate() 라는 메소드의 기능이 단순 totalCount 계산에서 자동으로 구매한 로또 티켓의 개수를 계산하도록 변경하게 되었습니다.
다음으로는 LottoMachine에서 purchase() 할 때 LottoPurchaseMoney 뿐 아니라 수동으로 생성한 List<LottoTicket> 을 함께 받도록 수정하게 되었습니다.
그러면서 테스트 코드도 다음과 같이 변경하게 되었는데요,
```java
@DisplayName("로또 티켓의 갯수를 반환한다.")
    @Test
    void checkLottoCount() {
        // given
        int lottoCount = 14;
        LottoTicketGenerator lottoTicketGenerator = new AutoLottoTicketGenerator();

        // when
        LottoTickets lottoTickets = new LottoTickets(lottoCount, new ArrayList<>(0), lottoTicketGenerator);

        // then
        assertThat(lottoTickets.totalCount()).isEqualTo(lottoCount);
    }
```
이 때 LottoTickets 의 생성자로 현재 `new ArrayList<>(0)` 을 전달해주고 있는데 이것이 적절한지는 의문이 듭니다...제가 생각할 때 현재 LottoTickets의 생성자를 직접 로 null이 전달되는 경우는 없기 때문에 null에 대한 검증을 하는 코드를 두지 않았는데, 테스트 코드에서 `new ArrayList<>(0)` 을 전달하는 부분이 마음에 걸립니다...

---

1단계 마지막 피드백에서 말씀해주신 다음의 말씀이 잘 이해가 가지 않습니다...
```
이 부분에서 고민이 필요해 보여요.
일단 1 ~ 45의 로또 숫자를 미리 생성해서 캐싱해놓는다는 관점에서는 잘 하셨어요 👍
다만 그렇다면 다음 사항들을 추가로 고려해서 리팩토링 해보면 좋을 듯 해요.

LottoNumber를 생성하려는 경우, 이미 생성되어 캐싱된 인스턴스가 있을 경우 새로 만들지 않고 기존 인스턴스를 반환하도록 한다.
혹은 LottoNumber 내부에서가 아니라, LottoNumber 생성을 전담하는 Pool을 별도로 만들어도 좋다.
캐싱하는 자료구조가 List인데, 더 나은 자료구조는 없을까?
```

LottoNumber를 생성하려는 경우, 이미 생성되어 캐싱된 인스턴스가 있을 경우 새로 만들지 않고 기존 인스턴스를 반환하도록 한다. -> 싱글톤 패턴의 적용이라고 이해하면 될까요..?? 하지만 LottoNumber 생성시 인자로 int 값을 받고 이를 상태로 가진다는 점에서 어떻게 싱글톤을 적용해야할지 모르겠습니다...ㅠㅠ

LottoNumber 내부에서가 아니라, LottoNumber 생성을 전담하는 Pool을 별도로 만들어도 좋다. -> 혹시 참고할만한 자료가 있을까요..Pool을 별도록 만든다는 말씀이 어떤 말씀이신지 모르겠습니다..

---

따끔한 피드백 부탁드립니다..!!

tco0427 committed:
refactor: InputView `getNormalWinningNumbers` 개선

replace()는 코드를 읽는 사람이 해석해야하는 단점이 존재하므로 split() 과 trim() 을 이용하도록 수정하였습니다.

tco0427 committed:
refactor: Rank 에서 사용하지 않는 필드 제거

Rank 열거형에서 isBonus 필드를 사용하지 않으므로 제거하였습니다. 불필요한 필드를 제거함으로써 Rank를 처음보는 사람도 가독성 있게 느낄 수 있습니다.

tco0427 committed:
refactor: LottoResult 메소드 추상화 정도 일정하게 변경

통계를 구하는 메소드에서 추상화 정도를 일정하게 변경해주었습니다.

tco0427 committed:
fix: 동일 당첨 시 제대로 계산 안되는 문제 해결

동일 등수(ex. 5등 2번) 당첨 시 제대로 계산이 되지 않는 문제를 해결하였습니다.

tco0427 committed:
refactor: 통계 출력 개선

통계 출력 부분에서 printf의 String.format을 이용하여 보다 가독성 있는 코드가 될 수 있도록 개선해보았습니다.

tco0427 committed:
refactor: 로또티켓 정보 출력문에 StringBuilder 적용

StringBuilder를 적용하여 문자열 합산 과정에서 속도나 저장공간 측면에서 이점을 가질 수 있도록 하였습니다.

tco0427 committed:
feat: 수동 구매 로또 수 입력 기능 추가

수동으로 로또 구매를 묻고 값을 얻어오는 InputView 기능 추가

tco0427 committed:
feat: 수동으로 구매할 번호를 묻는 안내문 추가

수동으로 구매할 로또의 번호를 묻는 안내문을 추가하였습니다.

tco0427 committed:
feat: 하나의 로또 티켓 번호 6개를 묻는 기능 추가

하나의 로또 티켓의 번호 6개를 묻는 기능을 추가하였습니다.

tco0427 committed:
feat: LottoTicket 의 생성자 오버로딩

수동 로또 번호를 통해서 LottoTicket 을 만들기 위한 생성자를 추가하였습니다.

tco0427 committed:
feat: LottoTickets 생성자 변경

LottoTickets 생성시 수동 로또 티켓을 함께 받도록 수정하였습니다.

tco0427 committed:
feat: LottoMachine 의 purchase 메소드에 수동 티켓 추가

LottoMachine 의 purchase 호출시 수동으로 생성한 LottoTicket 수를 통해서 자동으로 생성할 수를 계산하도록 기능을 개선하였습니다.

tco0427 committed:
feat: LottoController 수동 티켓 생성 기능 추가

수동 티켓 기능 추가에 따라 LottoController의 기능을 수정하였습니다.

tco0427 committed:
fix: 수동 티켓 생성에 따른 테스트 코드 수정

수동 티켓 생성에 따라 LottoMachine 및 LottoTicket 과 LottoPurchase 가 변경되었으므로 테스트 코드를 수정하였습니다.

tco0427 committed:
feat: 수동 로또 개수 및 숫자 입력 검증 추가

수동 로또 개수 검증 및 숫자 입력 검증하는 로직을 추가하였습니다.

tco0427 committed:
test: LottoPurchaseMoneyTest 수동 입력 개수 검증 추가

수동 입력 개수가 총 금액에 의한 개수보다 크면 안된다.

tco0427 committed:
docs(readme): LottoPurchaseMoney 요구사항 추가

수동 입력 개수가 총 금액에 의한 총 개수보다 큰지 검증한다.

tco0427 committed:
fix: 로또 구매 정보 출력문 수정

수동으로 몇개, 자동으로 몇개를 구매했는지를 표현해주기 위해 컨트롤러와 OutputView 를 수정하였습니다.

tco0427 committed:
docs(readme): 흐름도 수정

수동 로또 구매 기능을 포함함에 따라서 흐름도를 수정하였습니다.

wbluke commented on src/test/java/lotto/domain/LottoTicketsTest.java:
```diff
@@ -17,7 +18,7 @@ void checkLottoCount() {
 
         // when
-        LottoTickets lottoTickets = new LottoTickets(lottoCount, lottoTicketGenerator);
+        LottoTickets lottoTickets = new LottoTickets(lottoCount, new ArrayList<>(0), lottoTicketGenerator);
```
> 이 때 LottoTickets 의 생성자로 현재 `new ArrayList<>(0)` 을 전달해주고 있는데 이것이 적절한지는 의문이 듭니다...제가 생각할 때 현재 LottoTickets의 생성자를 직접 로 null이 전달되는 경우는 없기 때문에 null에 대한 검증을 하는 코드를 두지 않았는데, 테스트 코드에서 `new ArrayList<>(0)` 을 전달하는 부분이 마음에 걸립니다...

오호 어떤 이유에서 마음에 걸리는건지 궁금하네요.
수동 티켓 없이 totalCount()를 검증하고 싶다면 위와 같이 진행하면 되고, 수동 티켓도 포함해서 totalCount()를 검증하고 싶다면 티켓을 생성해서 넣어주면 되고요 ㅎㅎ
참고로 0 빼고 `new ArrayList<>()`만 해도 됩니다 :)

tco0427 replied:
위와 같은 형태의 테스트 코드를 처음 작성하다보니 어색하다는 느낌이 들었던 것 같아요..!! (null을 주는 경험이 더 많았다 보니 그런 것 같습니다.ㅎㅎ) new ArrayList<>() 의 생성자로 0을 준 이유는 initialCapacity가 0임을 명시적으로 하고 싶어서 0을 붙여줬습니다!!

wbluke commented on src/test/java/lotto/domain/LottoTicketsTest.java:
```diff
@@ -17,7 +18,7 @@ void checkLottoCount() {
+        LottoTickets lottoTickets = new LottoTickets(lottoCount, new ArrayList<>(0), lottoTicketGenerator);
 
         // then
         assertThat(lottoTickets.totalCount()).isEqualTo(lottoCount);
```
그리고 지금 같은 경우는 when절에 있는 LottoTickets 생성 까지가 given절이고,
lottoTickets.totalCount()가 when절이어야 합니다.
실수셨던거 같긴 한데, 항상 검증의 대상이 되는 행위가 when절에 위치해야 한다는 점 잊지마세요 ㅎㅎ
(그래서 보통 when절은 한줄로 끝나는 경우가 대부분입니다.)

다른 테스트도 비슷한 경우가 있는지 한번 체크해주세요 :)

tco0427 replied:
항상 given 과 when 의 경계가 모호하다는 생각을 했던 것 같아요..!! 그런데 닉의 `검증의 대상이 되는 행위`가 when절에 위치해야한다는 말씀을 들으니 조금 더 명확하게 이해가 됩니다..!! 다른 단위 테스트 부분들도 검토해보고 수정해보겠습니다!

wbluke commented on src/main/java/lotto/domain/LottoResult.java:
```diff
@@ -16,15 +17,33 @@ public LottoResult(Map<Rank, Integer> ranks) {
+            sum += rank.getPrizeMoney();
+        }
+        return sum;
+    }
```
음 외부에서 넣어준 sum에 더하기보다, 그냥 당첨횟수만큼 prizeMoney를 곱해서 반환하면 되지 않을까요? 
그럼 파라미터 수나 로직도 더 간단해질 듯 합니다 :)

tco0427 replied:
동의합니다..!! 메소드 추출(cmd + opt + m) 을 이용하다보니 코드에서 불필요한 매개변수등이 포함되게 된 것 같습니다..이런 단축키 사용 이후에 항상 한 번 더 확인하는 습관을 들여야하는데..앞으로 좀 더 주의하도록 하겠습니다!!

wbluke commented on src/main/java/lotto/domain/LottoResult.java:
```diff
@@ -16,15 +17,33 @@ public LottoResult(Map<Rank, Integer> ranks) {
+        }
+
+        return sum;
+    }
```
```java
int totalPrizeMoney = ranks.keySet()
        .stream()
        .mapToInt(Rank::getPrizeMoney)
        .sum();
```

엇 이 코드는 어디가고 다시 회귀했나요 ㅋㅋ

tco0427 replied:
맞는 말씀이십니다...ㅎㅎ 위의 getSum() 메소드를 수정하면서 getTotalPrizeMoney() 도 수정해보았습니다! 스트림을 활용하여 보다 가독성 있게 수정해보겠습니다!

wbluke commented on README.md:
```diff
@@ -61,7 +62,7 @@
+<img width="1115" alt="스크린샷 2022-03-02 오전 11 09 39" src="https://user-images.githubusercontent.com/57028386/156281285-09da0804-295e-4363-858b-c611e3c3c88d.png">
 
 <img width="983" alt="스크린샷 2022-02-24 오후 3 31 54" src="https://user-images.githubusercontent.com/57028386/155471018-9c88c6a1-2187-44ae-8582-523050c954e9.png">
 
```
(PR에 LottoNumber 변경사항이 없어서 여기에 남겨요)

> 1단계 마지막 피드백에서 말씀해주신 다음의 말씀이 잘 이해가 가지 않습니다...
> 
> ```md
> 이 부분에서 고민이 필요해 보여요.
> 일단 1 ~ 45의 로또 숫자를 미리 생성해서 캐싱해놓는다는 관점에서는 잘 하셨어요 👍
> 다만 그렇다면 다음 사항들을 추가로 고려해서 리팩토링 해보면 좋을 듯 해요.
> 
> LottoNumber를 생성하려는 경우, 이미 생성되어 캐싱된 인스턴스가 있을 경우 새로 만들지 않고 기존 인스턴스를 반환하도록 한다.
> 혹은 LottoNumber 내부에서가 아니라, LottoNumber 생성을 전담하는 Pool을 별도로 만들어도 좋다.
> 캐싱하는 자료구조가 List인데, 더 나은 자료구조는 없을까?
> ```
> 
> LottoNumber를 생성하려는 경우, 이미 생성되어 캐싱된 인스턴스가 있을 경우 새로 만들지 않고 기존 인스턴스를 반환하도록 한다. -> 싱글톤 패턴의 적용이라고 이해하면 될까요..?? 하지만 LottoNumber 생성시 인자로 int 값을 받고 이를 상태로 가진다는 점에서 어떻게 싱글톤을 적용해야할지 모르겠습니다...ㅠㅠ
> 
> LottoNumber 내부에서가 아니라, LottoNumber 생성을 전담하는 Pool을 별도로 만들어도 좋다. -> 혹시 참고할만한 자료가 있을까요..Pool을 별도록 만든다는 말씀이 어떤 말씀이신지 모르겠습니다..

아이구 제가 다른 분들 피드백과 같이 드리다보니 너무 집약적으로 이야기를 했나봅니다. 미안해요!
Pool이니 뭐니 이런 용어들은 일단 제외하고 본질적인 부분만 먼저 이야기를 하도록 할게요.

현재는 `new LottoNumber(1)` 하면 매번 새로운 인스턴스를 만들어주고 있어요.
그런데 과연 매번 새로운 인스턴스를 만들 필요가 있을까요?
```java
LottoNumber number1 = new LottoNumber(1);
LottoNumber number2 = new LottoNumber(1);
```
number1과 number2가 의미하는 것은 같은 로또숫자 1번인데, 인스턴스는 각각 생성되고 있어요.
그렇다면 **한번 생성된 인스턴스를 어딘가에 저장해두고, 다음번에 동일한 요청이 왔을 때 기존에 저장해놓은 인스턴스를 반환해주면 어떨까?** 라는 질문이에요. 
이를 캐싱이라고 표현할게요. (싱글턴과는 의미가 달라요 ㅎㅎ)

- 캐싱한 인스턴스를 어디에 보관하면 좋을까요? LottoNumber 자체에 static으로 보관해도 좋고, 별도의 클래스로 빼도 좋아요.
- 생성자는 말그대로 인스턴스를 만들어주는 도구라서 생성자 자체에서 해당 기능을 제공하기는 어려울 것이고, 정적 팩토리 메서드를 사용해 보세요.

고민해보시고 잘 모르겠으면 DM으로 말씀주세요. 추가 설명 드릴게요 :)

tco0427 replied:
천천히 읽어보니 이해가 되어 코드에 반영해보았습니다..!!

```java
    private LottoNumber(int number) {
        this.number = number;
    }

    public static LottoNumber getInstance(int number) {
        validateNumberBoundary(number);

        LottoNumber lottoNumber = LOTTO_TOTAL_NUMBERS.get(number - 1);

        return Objects.requireNonNullElseGet(lottoNumber, () -> new LottoNumber(number));
    }
```

또한 동일 인스턴스인지에 대한 검사를 진행하는 단위 테스트 코드를 LottoNumberTest 에 추가해보았습니다.

확인 부탁드립니다!!😃 감사합니다!!

wbluke reviewed (changes requested):
안녕하세요 디우~ 2단계도 잘 구현해 주셨네요 ㅎㅎ
질문 주신 부분 코멘트 다시 좀 더 상세히 남겼는데요, 확인 부탁드릴게요 🙂

tco0427 committed:
refactor: LottoResult 리팩토링

불필요한 매개변수 넘기는 부분을 제거하여 보다 간결하게 메소드를 리팩토링 하였습니다. 또한 stream을 활용하여 보다 간결하고 가독성 있는 코드가 되도록 하였습니다.

tco0427 committed:
refactor: 테스트 코드 개선

given 과 when 절의 구분이 불명확했던 테스트 코드를 찾아 수정하였습니다.
로그를 찍기 위해 사용했던 println() 구문을 제거하였습니다.

tco0427 committed:
fix: 통계 결과의 데이터 타입 long 으로 변경

int로 할 경우 1등 당첨이 2번일 시 수익률이 int의 최댓값을 넘어 마이너스로 출력되므로 이를 개선하기 위해 double형으로 계산하도록 수정하였습니다.

tco0427 committed:
fix: 수동으로 생성한 숫자도 정렬하도록 수정

수동으로 생성하는 로또 숫자 번호도 정렬하도록 수정하였습니다.

tco0427 committed:
fix: LottoResult 변경에 따른 Test 코드 수정

LottoResult 에서 소수점 아래 2자리에서 짜르지 않고 LottoResult 의 결과에서는 단순 double 형 값을 반환해주므로 이에 따라 수정을 진행하였습니다.

tco0427 committed:
fix: LottoNumber 동일 요청시 동일 인스턴스 반환

`LottoNumber` 생성에 대해서 동일 요청인 경우 동일 인스턴스를 반환하도록 정적 팩토리 메소드를 이용하여 적용해보았습니다.

tco0427 committed:
test: LottoNumber 동일 인스턴스 검증 추가

LottoNumber 에 대해서 동일 인스턴스인지를 검증하는 코드를 추가해주었습니다.
또한 LottoNumber 의 생성자를 private 으로 변경하고 정적 팩토리 메소드를 통해서 인스턴스를 반환받도록 수정함에 따라 영향을 받는 test 코드를 수정해주었습니다.

tco0427 committed:
refactor: LottoNumber 정적 팩토리 메소드 개선

Objects.requireNonNullElseGet() 을 사용하여 코드를 개선하엿습니다.
(Returns the first argument if it is non-null and otherwise returns the non-null value of supplier.get().)

wbluke commented on src/main/java/lotto/domain/LottoResult.java:
```diff
@@ -17,33 +17,32 @@ public LottoResult(Map<Rank, Integer> ranks) {
         }
-        return sum;
+
+        return result;
```
음 이거 지난번 피드백에서 더 간단하게 하기로 하지 않았나요?
```java
Integer count = ranks.get(rank);
return count * rank.getPrizeMoney();
```

tco0427 replied:
아하..위와 같은 말씀이셨군요..
`5000원 당첨 2번일 때 반복해야하네` 라는 생각으로 로직을 변경하면서 `반복`에 너무 집착해서 닉이 말씀해주신 내용을 제대로 이해하지 못한 것 같습니다...스스로를 반성하게 되는 피드백이었습니다! 감사합니다!!

wbluke commented on src/main/java/lotto/domain/vo/LottoNumber.java:
```diff
@@ -16,12 +16,19 @@
     }
 
-    private void validateNumberBoundary(int number) {
+    public static LottoNumber getInstance(int number) {
```
getInstance()는 보통 싱글턴 패턴에서 유일한 인스턴스를 반환할 때 사용하는 네이밍이고,
LottoNumber는 싱글턴이 아니니 적당한 정적 팩토리 메서드로 생각해서 네이밍해보면 좋겠어요. (ex. of())

tco0427 replied:
이펙티브 자바를 확인해보니 getInstance 라는 네이밍은 `매개변수로 명시한 인스턴스를 반환하지만 같은 인스턴스임을 보장하지는 않는다.` 라고 명시되어 있네요.. of는 `여러 매개변수를 받아 적합한 타입의 인스턴스를 반환하는 집계 메소드`라고 명시되어있습니다.
저는 개인적으로 `매개변수를 하나 받아서 해당 타입의 인스턴스를 반환하는 형변환 메소드` 를 의미하는 from이 가장 적절하다고 생각되어 해당 네이밍으로 수정해보도록 하겠습니다.

wbluke commented on src/main/java/lotto/domain/vo/LottoNumber.java:
```diff
@@ -16,12 +16,19 @@
+    public static LottoNumber getInstance(int number) {
+        validateNumberBoundary(number);
+
+        LottoNumber lottoNumber = LOTTO_TOTAL_NUMBERS.get(number - 1);
```
잘 하셨네요! 👍 
이제 좀 더 개선해 봅시다.
로또숫자를 의미하는 number가 List의 index 역할도 하고 있어요.
의미를 추가적으로 해석해야할 뿐더러, 로또숫자 리스트가 불연속적이거나 형태가 변경된다면 굉장히 위험할 수도 있는 코드입니다.
어떻게 하면 좋을까요? :)

힌트) 더 적절한 자료구조는 없을까?

tco0427 replied:
아하..!! 이전에는 더 적절한 자료구조가 없을까요? 라는 질문을 받았을 때 '왜 닉께서 이런 질문을 하셨을까...' 감조차 안잡혔던 것 같습니다.
그런데 직접 getInstance() 메소드를 구현하면서 LOTTO_TOTAL_NUMBERS.get() 을 사용하면서 스스로 '과연 적절한 코드일까?' 생각은 들었지만 우선은 그냥 사용했었는데요(코드에 적절한 이유가 없었습니다..ㅠㅠ 반성합니다..)
닉께서 말씀해주시는 `의미를 추가적으로 해석해야할 뿐더러, 로또숫자 리스트가 불연속적이거나 형태가 변경된다면 굉장히 위험할 수도 있는 코드입니다.` 다음 말씀을 듣고, Map 자료구조를 사용하는게 더욱 적절하겠다 라는 결론을 낼 수 있었습니다!! 감사합니다~_~

wbluke commented on src/test/java/lotto/domain/vo/LottoNumberTest.java:
```diff
@@ -10,15 +11,30 @@
+        LottoNumber otherInstance3 = LottoNumber.getInstance(2);
+
+        // when & then
+        assertThat(sameInstance1).isEqualTo(sameInstance2);
```
음 동일 인스턴스를 검증하기 위한 테스트인데, 검증 방법이 잘못되었어요.
한번 다시 고민해 보셔요 :)

tco0427 replied:
isEqualsTo 는 실제 값이 지정된 값과 같은지를 검증합니다. 따라서 말씀해주신대로 동일 인스턴스인지를 검증하는데는 적절하지 못한 것 같습니다...!!
isSameAs 와 isNotSameAs를 이용하면 제가 원하던 테스트를 진행할 수 있을 것 같습니다. (Verifies that the actual value is the same as the given one, ie using == comparison.)
수정하도록 하겠습니다.

wbluke reviewed (changes requested):
캐싱 전략 잘 적용해 주셨네요 ㅎㅎ
다만 아직 조금 더 개선할 부분이 보여요.
한번 더 확인 부탁드려요 🙂

tco0427 committed:
refactor: getSum 간략화

getSum() 메소드를 간략하게 수정하였습니다.

tco0427 committed:
refactor: 정적 팩토리 메소드 네이밍 변경

LottoNumber 의 정적 팩토리 메소드의 네이밍이 `매개변수를 하나 받아서 해당 타입의 인스턴스를 반환하는 형변환 메소드`를 의미하는 `from`이 보다 적절하다고 생각되어 변경하였습니다.

tco0427 committed:
fix: isSameAs 이용하여 동일 인스턴스 검증하도록 수정

isEqualTo 는 단순 내용을 비교하고, isSameAs는 주소값(==)을 비교하는 메소드이다.
따라서 동일 인스턴스 검증에는 isSameAs or isNotSameAs 를 이용하는 것이 보다 적절하여 수정하였습니다.

tco0427 committed:
refactor: LOTTO_TOTAL_NUMBERS 자료구조를 Map 으로 변경

이전에 로또 전체 번호(1~45)를 관리하기 위한 List는 인덱스를 사용하고, 다음과 같은 코드를 활용하였다.
`LottoNumber lottoNumber = LOTTO_TOTAL_NUMBERS.get(number - 1);`

하지만 이는 형태가 변경되면 매우 치명적일 수 있는 코드이므로 `Map` 자료구조가 더욱 적절하다고 판단되어 수정하였습니다.

wbluke commented on src/test/java/lotto/domain/vo/LottoNumberTest.java:
```diff
@@ -11,30 +11,30 @@
-        assertThat(sameInstance1).isEqualTo(sameInstance2);
-        assertThat(sameInstance1).isNotEqualTo(otherInstance3);
-        assertThat(sameInstance2).isNotEqualTo(otherInstance3);
+        assertThat(sameInstance1).isSameAs(sameInstance2);
```
잘 캐치하셨네요 ㅎㅎ
실제로 isEqualTo()와 isSameAs() 둘 다를 놓고, LottoNumber의 equals 메서드에 디버거 포인트 걸어놓고 돌리면 차이를 명확하게 확인하실 수 있으니 한번 해보셔요 :)

wbluke reviewed (approved):
로또 미션 고생 많으셨어요! 👍 
다음 미션도 화이팅입니다 🙂
